keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -71,6 +78,45 @@ class ResourceActuator(
 
         val response = vetoEnforcer.canCheck(resource)
         if (!response.allowed) {
+          /**
+           * [ArtifactVersioned] is a special [resource] sub-type. When a veto response sets
+           * [VetoResponse.vetoArtifact] and the resource under evaluation is of type
+           * [ArtifactVersioned], blacklist the desired artifact version from the environment
+           * containing [resource]. This ensures that the environment will be fully restored to
+           * a prior good-state.
+           */
+          if (response.vetoArtifact && resource.spec is ArtifactVersioned) {
+            try {
+              val (version, artifact) = when (desired) {
+                is Map<*, *> -> {
+                  if (desired.size > 0) {
+                    val versioned = (desired as Map<String, ArtifactVersioned>).values.first()
+                    Pair(versioned.artifactVersion, versioned.deliveryArtifact)
+                  } else {
+                    Pair(null, null)
+                  }
+                }
+                is ArtifactVersioned -> Pair(desired.artifactVersion, desired.deliveryArtifact)
+                else -> Pair(null, null)
+              }
+              if (version != null && artifact != null) {
+                val deliveryConfig = deliveryConfigRepository.deliveryConfigFor(resource.id)
+                val environment = deliveryConfig.environmentFor(resource)?.name
+                  ?: error("Failed to find environment for ${resource.id} in deliveryConfig ${deliveryConfig.name} " +
+                    "while attempting to veto artifact ${artifact.name} version $version")
+
+                artifactRepository.markAsVetoedIn(
+                  deliveryConfig = deliveryConfig,
+                  artifact = artifact,
+                  version = version,
+                  targetEnvironment = environment)
+                // TODO: emit event + metric
+              }
+            } catch (e: Exception) {
+              log.warn("Failed to veto presumed bad artifact version for ${resource.id}", e)
+              // TODO: emit metric
```

Comment:
Beyond metrics, I feel like this should cause a notification as it will have user impact.

Reply:
Not this PR.

Reply:
I guess my point was to add that to the `TODO` comment if it made sense. 😊